File: android/src/main/java/viska/android/ChatroomActivity.java
package viska.android;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.couchbase.lite.Document;
import com.couchbase.lite.ListenerToken;
import com.google.android.material.appbar.MaterialToolbar;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import viska.database.Chatroom;

public class ChatroomActivity extends InstanceActivity {

  /**
   * Starts this activity.
   *
   * @param chatroomMembers Account ID of the chatroom members.
   */
  public static void start(final Context source, final Collection<String> chatroomMembers) {
    // Will be like viska://chatroom/account1+account2+...
    final Uri uri =
        new Uri.Builder()
            .scheme("viska")
            .authority("chatroom")
            .appendPath(String.join("+", chatroomMembers))
            .build();
    final Intent intent = new Intent(source, ChatroomActivity.class);
    intent.setData(uri);
    source.startActivity(intent);
  }

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.chatroom);

    final RecyclerView list = findViewById(R.id.list);
    list.setLayoutManager(new LinearLayoutManager(this));
  }

  @Override
  protected void onStart() {
    super.onStart();

    final MaterialToolbar actionBar = findViewById(R.id.action_bar);
    setSupportActionBar(actionBar);

    final List<String> chatroomMembers = getChatroomMembers();
    if (chatroomMembers.isEmpty()) {
      return;
    }

    final String chatroomId = Chatroom.Companion.getChatroomIdFromMembers(chatroomMembers);
    final ListenerToken token =
        db.addDocumentChangeListener(
            Chatroom.Companion.getDocumentId(chatroomId),
            change -> {
              final Document document = change.getDatabase().getDocument(change.getDocumentID());
              if (document != null) {
                final Chatroom chatroom = new Chatroom(change.getDatabase(), document);
                setTitle(chatroom.getDisplayName());
              }
            });
    storeListenerToken(token);

    final RecyclerView list = findViewById(R.id.list);
    final ConversationAdapter adapter = new ConversationAdapter(db, chatroomMembers);
    list.setAdapter(adapter);
  }

  @Override
  protected void onStop() {
    final RecyclerView list = findViewById(R.id.list);
    ((CouchbaseLiveQueryListAdapter) list.getAdapter()).unsubscribe();

    super.onStop();
  }

  private List<String> getChatroomMembers() {
    final Uri uri = getIntent().getData();
    return uri == null
        ? Collections.emptyList()
        : Arrays.asList(uri.getLastPathSegment().split("\\+"));
  }
}

File: android/src/main/java/viska/android/ChatroomActivity.kt
package viska.android

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.preferredSize
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material.ListItem
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Scaffold
import androidx.compose.material.Text
import androidx.compose.material.TopAppBar
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.setContent
import androidx.compose.ui.unit.dp
import com.google.protobuf.ByteString
import com.google.protobuf.BytesValue
import dagger.Lazy
import dagger.hilt.android.AndroidEntryPoint
import java.lang.IllegalArgumentException
import javax.inject.Inject
import viska.daemon.Daemon.Message
import viska.database.displayId
import viska.database.toBinaryId

@AndroidEntryPoint
class ChatroomActivity : InstanceActivity() {

  @Inject lateinit var daemonService: Lazy<viska.daemon.DaemonService>

  override fun onCreate(savedInstanceState: Bundle?) {
    try {
      super.onCreate(savedInstanceState)
    } catch (_: ActivityRedirectedException) {
      return
    }

    setContent {
      MaterialTheme {
        val chatroomIdProtobuf = BytesValue.of(ByteString.copyFrom(chatroomId.toBinaryId()))
        val chatroom by daemonService
            .get()
            .nodeGrpcClient
            .watchChatroom(chatroomIdProtobuf)
            .collectAsState(null)
        if (chatroom == null) {
          finish()
          return@MaterialTheme
        }

        val messagesSubscription by daemonService
            .get()
            .nodeGrpcClient
            .watchChatroomMessages(chatroomIdProtobuf)
            .collectAsState(null)

        Scaffold(topBar = { TopAppBar(title = { Text(text = chatroom?.name ?: "") }) }) { _ ->
          LazyColumn {
            items(messagesSubscription?.messagesList ?: emptyList()) { MessageItem(it) }
          }
        }
      }
    }
  }

  private val chatroomId: String
    get() {
      val uriPath = intent.data?.pathSegments ?: emptyList()
      if (uriPath.size != 2 && "chatroom" != uriPath[0]) {
        throw IllegalArgumentException("Bad chatroom URI")
      }
      return uriPath[1]
    }

  companion object {

    fun start(source: Context, chatroomId: ByteArray) {
      // Will be like:
      // viska://chatroom/87956192A8143476909113CDA0D4077E092E26E10CC7DAC43E68F694EA68A036
      val uri =
          Uri.Builder()
              .scheme("viska")
              .authority("chatroom")
              .appendPath(chatroomId.displayId())
              .build()
      val intent = Intent(source, ChatroomActivity::class.java)
      intent.data = uri
      source.startActivity(intent)
    }
  }
}

@Composable
private fun MessageItem(message: Message) {
  ListItem(
      icon = { Image(imageVector = Icons.Default.Person, Modifier.preferredSize(48.dp)) },
      text = { Text(text = message.content) })
}
